src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsCache.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.UnifiedSuggestions;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
+{
+    /// <summary>
+    /// Caches suggested action sets between calls to <see cref="CodeActionsHandler"/> and
+    /// <see cref="CodeActionResolveHandler"/>.
+    /// </summary>
+    [Export(typeof(CodeActionsCache)), Shared]
+    internal class CodeActionsCache
+    {
+        /// <summary>
+        /// Ensures we aren't making concurrent modifications to the list of cached items.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        /// <summary>
+        /// Maximum number of cached items.
+        /// </summary>
+        private readonly int _maxCacheSize = 3;
+
+        /// <summary>
+        /// Current list of cached items.
+        /// </summary>
+        private readonly List<CodeActionsCacheItem> _cachedItems = new List<CodeActionsCacheItem>();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionsCache()
+        {
+        }
+
+        public async Task UpdateCacheAsync(
```

Comment:
Nit: `cache.GetCacheAsync()` sounds weird to me. Something like `cache.GetCodeActionsAsync()` is what I would expect. Applies to the Update method too